### PR TITLE
Cleanup return_text and return_full_text options in TextGenerationPipeline

### LINKED ISSUE #33535: TextGenerationPipeline: ValueError: `return_text` is mutually exclusive with `return_full_text`
### System Info

- `transformers` version: 4.44.2
- Platform: Linux-6.8.0-40-generic-x86_64-with-glibc2.35
- Python version: 3.11.4
- Huggingface_hub version: 0.24.6
- Safetensors version: 0.4.5
- Accelerate version: 0.24.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.0+cu117 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: no

### Who can help?

@Rocketknight1 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```python
from transformers.pipelines import pipeline, TextGenerationPipeline
template = """
<s>
{%- for message in messages %}
    {{- message['role'] + ' ' + message['content'] }}
{%- endfor %}
    assistant
"""

model_name = "TurkuNLP/gpt3-finnish-small"
model_pipeline: TextGenerationPipeline = pipeline(model=model_name, framework="pt", task="text-generation")
model_pipeline.tokenizer.chat_template = template

messages = [{"role": "user", "content": "Mitä Suomen paras kaupunki on?"}]
output = model_pipeline(messages, return_text=True, return_full_text=False, do_sample=True, max_new_tokens=10, num_return_sequences=2)
print(output)
```

(Nevermind that this template for this model doesn't really make sense).
The above raises
```
File .../transformers/pipelines/text_generation.py:174, in TextGenerationPipeline._sanitize_parameters(self, return_full_text, return_tensors, return_text, return_type, clean_up_tokenization_spaces, prefix, handle_long_generation, stop_sequence, truncation, max_length, **generate_kwargs)
    172 if return_full_text is not None and return_type is None:
    173     if return_text is not None:
--> 174         raise ValueError("`return_text` is mutually exclusive with `return_full_text`")
    175     if return_tensors is not None:
    176         raise ValueError("`return_full_text` is mutually exclusive with `return_tensors`")

ValueError: `return_text` is mutually exclusive with `return_full_text`
```

Removing the `return_text=True` argument to the pipeline call returns the expected result (only the newly generated text).

### Expected behavior

The code above succeeds and returns only newly generated text, without the prompt, as specified in [the documentation](https://huggingface.co/docs/transformers/main/en/main_classes/pipelines#transformers.TextGenerationPipeline.__call__): 

> return_text (bool, optional, defaults to True) — Whether or not to return the decoded texts in the outputs.
> return_full_text (bool, optional, defaults to True) — If set to False only added text is returned, otherwise the full text is returned. Only meaningful if return_text is set to True.

### PR DESCRIPTION
**Pull request by Rocketknight1**
The extra options in `TextGenerationPipeline` right now are confusing and poorly documented. This PR refactors things so that the documentation is clearer. It also creates a proper hierarchy of options (`return_full_text` overrides and ignores `return_text`, and `return_tensors` is incompatible with both)

Before this PR, the documentation was incorrect, as it claimed that `return_full_text` required `return_text`, whereas actually these options were not compatible with each other!

Behaviour after the change should be mostly unchanged, with the exception that `return_full_text` overrides `return_text` if both are set, rather than raising an error. The test that checked for those two options being incompatible has been removed.

Fixes #33535

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @LysandreJik for review!

**Comment by Rocketknight1** [comment]
Maybe cc @amyeroberts actually, since Lysandre is still away (also I know this is 6 on a Friday, don't worry about reviewing it until Monday!)

File: src/transformers/pipelines/text_generation.py [hunk 1]
```diff
@@ -173,16 +173,21 @@ def _sanitize_parameters(
...
+            )
+        if return_tensors is not None:
+            if return_text or return_full_text:
+                raise ValueError("`return_text` and `return_full_text` are mutually exclusive with `return_tensors`")
             return_type = ReturnType.TENSORS
+        elif return_full_text and return_type is None:
+            return_type = ReturnType.FULL_TEXT
+        elif (return_text or return_full_text is not None) and return_type is None:
+            # Explicitly setting return_full_text = False is caught here, and results in ReturnType.NEW_TEXT
+            return_type = ReturnType.NEW_TEXT
```
Comment (amyeroberts) [inline src/transformers/pipelines/text_generation.py:190]: I'm confused why we would have `ReturnType.NEW_TEXT` if `return_text=None` and `return_full_text=False`

File: tests/pipelines/test_pipelines_text_generation.py [hunk 2]
```diff
@@ -435,8 +435,6 @@ def run_pipeline_test(self, text_generator, _):
                 ],
             )
 
-        with self.assertRaises(ValueError):
-            outputs = text_generator("test", return_full_text=True, return_text=True)
```
Comment (amyeroberts) [inline tests/pipelines/test_pipelines_text_generation.py:439]: I'm also not sure why we remove a check that they're mutually exclusive -- I think it's far better to raise an exception if both are set rather than have one silently override the other

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

Two questions about the intended behaviour. It might be that I'm not familiar with the way the pipeline is normally used

File: src/transformers/pipelines/text_generation.py (hunk 1 above)
Comment (LysandreJik) [inline src/transformers/pipelines/text_generation.py:190]: This whole block is really hard to parse, is it possible to add white space/comments?

File: tests/pipelines/test_pipelines_text_generation.py (hunk 2 above)
Comment (Rocketknight1) [reply tests/pipelines/test_pipelines_text_generation.py:439]: I was trying to reconcile the docs (which used to be very wrong) with the options, but you're right - I'll make them mutually exclusive like they were before, and update the docs instead.

**Comment by Rocketknight1** [comment]
@amyeroberts @LysandreJik I reworked the PR in response to Amy's comment - I think she was right about the new behaviour not being any better. I restored the tests and behaviour to how they were before, and instead updated the docs so they correctly reflect that behaviour. As a result, this PR only touches docstrings now!

**Review by LysandreJik — APPROVED** [review]
That's certainly easier to review :smile:

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating and fixing the docs!
